appsrc/dcfile/dctopdf.man
.TH DCTOPDF 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - Extract DICOM encapsulated PDF"
.SH NAME
dctopdf \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Extract DICOM encapsulated PDF
.SH SYNOPSIS
.HP 10
.B dctopdf
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-quiet|silent
]
.so man1/optin.so
.so man1/binout.so
.SH DESCRIPTION
.LP
.B dctopdf
reads the named dicom input file containing an encpasulated PDF file and stores it as a PDF file.
.LP
.SH OPTIONS
The PDF output goes where it is specified or standard out.
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-quiet|silent
.RS
Suppress the normal description of encapsulated document parameters.
.RE
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR pdftodc(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/dcfile/dctopdf.cc
static const char *CopyrightIdentifier(void) { return "@(#)dctopdf.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "attrothr.h"
#include "mesgtext.h"
#include "ioopt.h"
#include "dcopt.h"
#include "elmconst.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	OutputOptions		output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool quiet=options.get("quiet") || options.get("silent");

	dicom_input_options.done();
	output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	OutputOpenerFromOptions output_opener(
		options,output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< output_options.usage()
			<< " [-v|-verbose]"
			<< " [-quiet|-silent]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	ostream out(output_opener);
	BinaryOutputStream bout(out,NoEndian);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	Attribute *apdfdata=list[TagFromName(EncapsulatedDocument)];
	if (!apdfdata) {
		log << EMsgDC(MissingAttribute) << " - \"EncapsulatedDocument\"" << endl;
		success=false;
	}
	else if (!apdfdata->isOtherByteNonPixel()) {
		log << EMsgDC(EncapsulatedDataIncorrectVR) << endl;
		success=false;
	}
	else {
		if (!quiet) {
			log << "******** Parameters ... ********" << endl; 
			ElementDictionary *dict=list.getDictionary();
			Attribute *a;
			a=list[TagFromName(MIMETypeOfEncapsulatedDocument)];
			if (a) { a->write(log,dict); log << endl; }
		}

		apdfdata->writeData(bout);

		if (!bout.good()) {
			log << EMsgDC(Writefailed) << endl;
			success=false;
		}
	}

	return success ? 0 : 1;
}

	
